### ISSUE TITLE
Circular References Take 2

### ISSUE THREAD

**Issue by rozza** [hmarr/mongoengine#312]
@jpfarias still having issues with #294

Add tests for:

Using lambda for default:

```
a = EmbeddedDocumentField(A, default: lambda A())
```

Making a cyclic tree like with more than 2 objects is a better test:

```
 a -> b -> c -> a
```

Another more complex test would be: 

```
  a -> [b, c, d]
  b -> [a, c, d]
  c -> [a, b, d]
  d -> [a, b, c]
```

**Comment by jpfarias** [comment]
Test that fails, to be added on tests/dereference.py:

```
    def test_circular_tree_reference(self):
        """Ensure you can handle circular references with more than one level
        """
        class Other(EmbeddedDocument):
            name = StringField()
            friends = ListField(ReferenceField('Person'))

        class Person(Document):
            name = StringField()
            other = EmbeddedDocumentField(Other, default=lambda: Other())
            def __repr__(self):
                return "<Person: %s>" % self.name


        Person.drop_collection()
        paul = Person(name="Paul"); paul.save()
        maria = Person(name="Maria"); maria.save()
        julia = Person(name='Julia'); julia.save()
        anna = Person(name='Anna'); anna.save()

        paul.other.friends = [maria, julia, anna]
        paul.other.name = "Paul's friends"
        paul.save()

        maria.other.friends = [paul, julia, anna]
        maria.other.name = "Maria's friends"
        maria.save()

        julia.other.friends = [paul, maria, anna]
        julia.other.name = "Julia's friends"
        julia.save()

        anna.other.friends = [paul, maria, julia]
        anna.other.name = "Anna's friends"
        anna.save()

        self.assertEquals(
            "[<Person: Paul>, <Person: Maria>, <Person: Julia>, <Person: Anna>]", 
            "%s" % Person.objects()
        )

```

Fails with: RuntimeError: maximum recursion depth exceeded while calling a Python object

**Comment by jpfarias** [comment]
Changed the _get_changed_fields() to keep track of things it already inspected so it doesn't need to inspect them again, seems to work for me:

```
    def _get_changed_fields(self, key='', inspected=None):
        """Returns a list of all fields that have explicitly been changed.
        """
        from mongoengine import EmbeddedDocument
        _changed_fields = []
        _changed_fields += getattr(self, '_changed_fields', [])
        if inspected is None:
            inspected = set()
        if hasattr(self, 'id'):
            if self.id in inspected:
                return _changed_fields
            inspected.add(self.id)

        for field_name in self._fields:
            db_field_name = self._db_field_map.get(field_name, field_name)
            key = '%s.' % db_field_name
            field = getattr(self, field_name, None)
            if hasattr(field, 'id'):
                if field.id in inspected:
                    continue
                inspected.add(field.id)

            if isinstance(field, EmbeddedDocument) and db_field_name not in _changed_fields:  # Grab all embedded fields that have been changed
                _changed_fields += ["%s%s" % (key, k) for k in field._get_changed_fields(key, inspected) if k]
            elif isinstance(field, (list, tuple, dict)) and db_field_name not in _changed_fields:  # Loop list / dict fields as they contain documents
                # Determine the iterator to use
                if not hasattr(field, 'items'):
                    iterator = enumerate(field)
                else:
                    iterator = field.iteritems()
                for index, value in iterator:
                    if not hasattr(value, '_get_changed_fields'):
                        continue
                    list_key = "%s%s." % (key, index)
                    _changed_fields += ["%s%s" % (list_key, k) for k in value._get_changed_fields(list_key, inspected) if k]

        return _changed_fields
```

**Comment by rozza** [comment]
Fixed in dev: 7b1860d17b0d60bd11aaace9052bd58a2d702cbf

Thanks again to @jpfarias
